rcl_executor/src/let_executor.c

```diff
@@ -0,0 +1,597 @@
+      e->allocator->state);
+  if (NULL == e->handles) {
+    RCL_SET_ERROR_MSG("Could not allocate memory for 'handles'.");
+    return RCL_RET_BAD_ALLOC;
+  }
+
+  // initialize handle
+  for (size_t i = 0; i < number_of_handles; i++) {
+    rcle_handle_init(&e->handles[i], number_of_handles);
+  }
+
+  // debug: output handle
+  /*
+  for (size_t i = 0; i < number_of_handles;i++)
+  {
+          rcle_handle_print(& e->handles[i] );
+  }
+  */
+  // initialize #counts for handle types
+  rcle_handle_size_zero_init(&e->info);
+  e->initialized = true;
+  return ret;
+}
+
+rcl_ret_t
+rcle_let_executor_set_timeout(rcle_let_executor_t * executor, const uint64_t timeout_ns)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    executor, "executor is null pointer", return RCL_RET_INVALID_ARGUMENT);
+  rcl_ret_t ret = RCL_RET_OK;
+  if (executor->initialized) {
+    executor->timeout_ns = timeout_ns;
+  } else {
+    RCL_SET_ERROR_MSG("executor not initialized.");
+    return RCL_RET_ERROR;
+  }
+  return ret;
+}
+
+rcl_ret_t
+rcle_let_executor_fini(rcle_let_executor_t * executor)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    executor, "executor is null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    executor->handles, "memory for exector->handles is NULL", return RCL_RET_ERROR);
+  rcl_ret_t ret = RCL_RET_OK;
+
+  if (executor->initialized) {
+    executor->allocator->deallocate(executor->handles, executor->allocator->state);
+    executor->max_handles = 0;
+    executor->index = 0;
+    rcle_handle_size_zero_init(&executor->info);
+
+    // free memory of wait_set if it has been initialized
+    // calling it with un-initialized wait_set will fail.
+    if (executor->wait_set_initialized) {
+      rcl_ret_t rc = rcl_wait_set_fini(&executor->wait_set);
+      if (rc != RCL_RET_OK) {
+        PRINT_RCL_ERROR(rcle_let_executor_fini, rcl_wait_set_fini);
+      }
+      executor->wait_set_initialized = false;
+    }
+
+    executor->initialized = false;
+    // todo(jst3si) reset timeout to default value
+  } else {
+    RCL_SET_ERROR_MSG("executor not initialized or called _fini function twice");
+    ret = RCL_RET_ERROR;     // TODO(jst3si) better name for "calling this function multiple times"
+  }
+  return ret;
+}
+
+
+rcl_ret_t
+rcle_let_executor_add_subscription(
+  rcle_let_executor_t * executor,
+  rcl_subscription_t * subscription,
+  void * msg,
+  rcle_callback_t callback,
+  rcle_invocation_t invocation)
+{
+  rcl_ret_t ret = RCL_RET_OK;
+
+  RCL_CHECK_ARGUMENT_FOR_NULL(executor, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(subscription, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(msg, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(callback, RCL_RET_INVALID_ARGUMENT);
+
+  // array bound check
+  if (executor->index >= executor->max_handles) {
+    rcl_ret_t ret = RCL_RET_ERROR;     // TODO(jst3si) better name : RCLE_RET_BUFFER_OVERFLOW
```

Comment:
no need to local scope variable just in here.
```suggestion
    ret = RCL_RET_ERROR;     // TODO(jst3si) better name : RCLE_RET_BUFFER_OVERFLOW
```

Reply:
Resolved in cdbfd4d16ca125d034295ceb4a0e5921158d4f0e